### ISSUE TITLE
Lazy object instantiation via config file

### ISSUE THREAD

**Issue by rob-hen** [Lightning-AI/pytorch-lightning#19270]
### Description & Motivation

As far as I can see, there is no way currently to do lazy instantiation.

Let say, we want to create an object of class `path.to.B` that we configure from the yaml file. However, its creation might depend on a property of another object, that we know only during runtime. So it would be great if, based on the config file, we get a constructor for an object of class `path.to.B` , where all parameters are already set from the config file. In the example below, the constructor would already set `u=1, v=2`, e.g. using `functools.partial`. So instead of getting the instantiated object I get a preconfigured constructor. 

See the example below

Module:

```python
class Model(pl.LightningModule):

     def __init__(self,model):
         super().__init__()
        obj_A = self.model.A
        obj_B_constructor = self.model.B
        obj_B = obj_B_constructor(extern_param=obj_A.param) # should correspond to path.to.B_class(u=1, v=2, extern_param=obj_A)
``` 
config file:
```
model:
    A:
      class_path: path.to.A_class
      init_args:
           a: 1
           b: 2
    B:
     class_path: path.to.B_class
     lazy_init: True
     init_args:
         u: 1
         v: 2
``` 

### Pitch

By setting an additional key in the config file, we are not getting the instantiate object, but the constructor for the object, with all parameters already set.

**Comment by awaelchli** [comment]
@rob-hen I assume you are talking about LightningCLI here? There is a lazy_instance method, explained here at the end of the section: https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_expert.html#class-type-defaults

This is the closest to lazy instantiation we have, that I know of.

**Comment by rob-hen** [comment]
@awaelchli Yes, I am using the LightningCLI e.g. `python train.py --config config.yaml`. I would like to define lazy init with the config file, with the effect as described above.

**Comment by mauvilsa** [comment]
If the number of "external" parameters is fixed, and they are the first positional parameters, then it is the same case as the optimizers and schedulers, see [multiple-optimizers-and-schedulers](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_advanced_3.html#multiple-optimizers-and-schedulers). That is, you would define the parameter and type as `model_b_constructor: Callable[[TypeExternalParam], BaseClassB]`. And then in the `__init__` body do `obj_B = model_b_constructor(obj_A.param)`.

The proposed `lazy_init: True` pattern will not be done. It is a design decision that type hints are strictly followed. If someone did `model_b: BaseClassB` and then does lazy init, then the `__init__` would not get an instance of `BaseClassB`. This is conceptually wrong. Also, in the code always an extra parameter needs to be given to create the instance. So it shouldn't be up to the CLI user to give in a config `lazy_init: True`. This should be part of the python code, not configuration.

**Comment by rob-hen** [comment]
@mauvilsa How would the corresponding config file look like? When I use `Callable`, it was already creating the object.

**Comment by rob-hen** [comment]
`Not sure what you mean by "it was already creating the object".`: My object `model.constructor` was not the `__init__` function, but the instantiated object.


I created an MWE to show what I want: Lets say we have the following file `test.py`:

```python
import pytorch_lightning as pl
from pytorch_lightning.cli import LightningCLI
from pytorch_lightning.demos.boring_classes import BoringDataModule
from test_classes import ModelParams


class Model(pl.LightningModule):

    def __init__(self, model_params: ModelParams):
        super().__init__()
        arg1 = 1
        arg2 = 2
        B_constructor = model_params.constructor
        obj = B_constructor(arg1, arg2)
        print(obj.arg1 == arg1)
        print(obj.arg2 == arg2)
        print(obj.paramA == 3)


if __name__ == "__main__":
    cli = LightningCLI(Model, BoringDataModule)

``` 
and `test_classes.py`:
```python
from typing import Callable


class AbstractB:

    def __init__(self, arg1, arg2):
        self.arg1 = arg1
        self.arg2 = arg2


class B(AbstractB):

    def __init__(self, arg1, arg2, paramA):
        super().__init__(arg1, arg2)
        self.paramA = paramA


class ModelParams():
    def __init__(self, constructor: Callable[[int, int], B]):
        self.constructor = constructor

``` 
together with the config file `test.yaml`:
```
model:
  model_params:
    class_path: test_classes.ModelParams
    init_args:
      constructor:
        class_path: test_classes.AbstractB
        init_args:
          paramA: 3
``` 
Now `python test.py fit --config test.yaml` raise an error: `Type typing.Callable[[int, int], test_classes.B] expects a function or a callable class: Expected 'test_classes.AbstractB' to be a class that instantiates into callable or a subclass of False. Got value: Namespace(class_path='test_classes.AbstractB', init_args=Namespace(paramA=3))`

What do I need to do in order to use `Callable`?

**Comment by mauvilsa** [comment]
The code and config above works as expected to me. Not sure what you have done differently. Do you have the latest version of jsonargparse installed?

I added `print(f"constructor: {constructor}")` to `ModelParams` and added `print(f"B_constructor: {B_constructor}")`, `print(f"obj: {obj}")` to `Model`. With these additions the output is
```shell
$ python3 test.py fit --config test.yaml 
...
constructor: <function adapt_class_type.<locals>.partial_instance at 0x7fcbdb36f2e0>
B_constructor: <function adapt_class_type.<locals>.partial_instance at 0x7fcbdb36f2e0>
obj: <test_classes.B object at 0x7fcbdb7e3a60>
True
True
True
...
```
Exactly as expected.


I tried a few other modifications to make it fail, but I was unable to reproduce the exact same error. The only way I got an error was by redefining `B` as:
```python
class B(AbstractB):
    def __init__(self, *args, paramA: int):
        super().__init__(*args)
        self.paramA = paramA
```
I would have hoped that this works also, but there seems to be an issue there. Anyway, the error is different.

By the way, please add type hints. Without type hints it doesn't fail for me. But the CLI is designed to work best when there are types.

**Comment by rob-hen** [comment]
Indeed, a had an old jsonargparse version (`4.20.1`). With the latest version it works correctly.

Now I am wondering, if the approach can still be improved?
The one direction is not to limit the number of positional arguments (with `*args`  as you were showing it). What would be the corresponding code for 
```python
class ModelParams():
    def __init__(self, constructor: Callable[[int, int], AbstractB]):
        self.constructor = constructor
``` 
?
Would it be 
```python
class ModelParams():
    def __init__(self, constructor: Callable[[List[Any]], AbstractB]):
        self.constructor = constructor
``` 
Can we get do something like `bj = B_constructor(arg1=arg1, arg2=arg1)`?

**Comment by e-dorigatti** [comment]
I have a similar issue, namely that to instantiate the model we need to know the shape of the data, which is only known after downloading and preparing the data. I am currently solving it this way:


    def main():
        cli = LightningCLI(auto_configure_optimizers=False, run=False, ...)
        dm, mod = cli.datamodule, cli.model

        dm.prepare_data()
        dm.setup('fit')

        mod.initialize_modules(dm.input_shape, dm.output_shape)

        cli.trainer.fit(mod, dm) 

    class DataModule(LightningDataModule):
        # etc.

        def setup(self, stage):
            self.x, self.y = self.load_data()
            self.input_shape = self.x.shape[1]
            self.output_shape = self.y.shape[1]

    class TrainingModule(LightningModule):
        # etc.

        def initialize_modules(self, input_shape, output_shape):
            self.predictor = nn.Sequential(
                nn.Linear(input_shape, self.hidden_shape), nn.ReLU(),
                nn.Linear(self.hidden_shape, output_shape)
            )

        def configure_optimizers(self):
            return ...
        

The disadvantage is that in this way I would need to implement all other commands (`validate`, `test`, `predict`).

Is there a more canonical/standard way to achieve the same result? If not, I think this is a valid use-case to support (in any case, this saves a lot of code compared to not using lightning, so thank you a lot already for this gem!)

**Comment by mauvilsa** [comment]
> Is there a more canonical/standard way to achieve the same result? If not, I think this is a valid use-case to support (in any case, this saves a lot of code compared to not using lightning, so thank you a lot already for this gem!)

This is a use case for links applied on instantiation, see [cli docs](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_expert.html#argument-linking) or [jsonargparse docs](https://jsonargparse.readthedocs.io/en/v4.27.2/#applied-on-instantiate).

One possibility would be to implement in the LightningDataModule module two `@property` methods for the shapes. Accessing the property should automatically run whatever is needed to compute the shapes (i.e. setup). The LightningModule should receive in the`__init__` the shapes. Finally create instantiation links from the data properties to the module parameters.

Having the shape dependent on the data is somewhat tricky. When you run `fit` it can make sense to compute the shapes from the training data. But if you later want to test, a checkpoint would be loaded and the shape be implied by that. It it is possible to compute the same shape from the test data, then it can work without issue, though it is unnecessary to compute it since the model comes from the checkpoint.

**Comment by mauvilsa** [comment]
> With a `Protocol` it is possible to define a `__call__` with variable positionals. Though, the previous limitations remain.

With the latest version of jsonargparse it is now possible to use a callable protocol, see [instance-factories](https://jsonargparse.readthedocs.io/en/v4.41.0/#instance-factories). Though, it was implemented thinking more on allowing to call with keyword arguments and not be limited by order of positionals.
